database/src/main/java/com/firebase/ui/database/FirebaseArrayOfObjects.java

```diff
@@ -0,0 +1,237 @@
+                                                            SnapshotParser<T> parser) {
+        FirebaseArrayOfObjects<T> array = newInstance(snapshots, modelClass);
+        array.mParser = parser;
+        return array;
+    }
+
+    public List<DataSnapshot> getSnapshots() {
+        return mSnapshots;
+    }
+
+    protected List<E> getObjects() {
+        List<E> objects = new ArrayList<>(mSnapshots.size());
+        for (int i = 0; i < mSnapshots.size(); i++) {
+            objects.add(get(i));
+        }
+        return objects;
+    }
+
+    @Override
+    public int size() {
+        return mSnapshots.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return mSnapshots.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return indexOf(o) >= 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable iterator
+     */
+    @Override
+    public Iterator<E> iterator() {
+        return new ImmutableIterator(getObjects().iterator());
+    }
+
+    @Override
+    public Object[] toArray() {
+        return getObjects().toArray();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        return getObjects().toArray(a);
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        return getObjects().containsAll(c);
+    }
+
+    @Override
+    public E get(int index) {
+        return mParser.parseSnapshot(mSnapshots.get(index));
+    }
+
+    @Override
+    public int indexOf(Object o) {
+        return getObjects().indexOf(o);
+    }
+
+    @Override
+    public int lastIndexOf(Object o) {
+        return getObjects().lastIndexOf(o);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable list iterator
+     */
+    @Override
+    public ListIterator<E> listIterator() {
+        return new ImmutableListIterator(getObjects().listIterator());
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable list iterator
+     */
+    @Override
+    public ListIterator<E> listIterator(int index) {
+        return new ImmutableListIterator(getObjects().listIterator(index));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+
+        FirebaseArrayOfObjects<?> array = (FirebaseArrayOfObjects<?>) obj;
+
+        return mSnapshots.equals(array.mSnapshots) && mEClass.equals(array.mEClass);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = mSnapshots.hashCode();
+        result = 31 * result + mEClass.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "FirebaseArrayOfObjects{" +
+                "mSnapshots=" + mSnapshots +
+                '}';
+    }
+
+    protected static class FirebaseArrayOfObjectsOptimized<E> extends FirebaseArrayOfObjects<E>
```

Comment:
@samtstern I said I want people to be able to extend it, not the opposite. Here's what I would like:

- People **can** extend `FirebaseArrayOfObjectsOptimized`
- People **cannot** create a new instance from outside `FirebaseArrayOfObjects`

@samtstern as for making a normal class `protected`, you can't do that. It goes `public` -> `default` and that's it. For people to be able to extend `FirebaseArrayOfObjectsOptimized` the class would need to be public, but that makes a confusing api because people will wonder what the difference between `FirebaseArrayOfObjects` and `FirebaseArrayOfObjectsOptimized` is.

@samtstern Given what I said above, do you still think `FirebaseArrayOfObjectsOptimized` should still be in `FirebaseArrayOfObjects`? I know it makes it confusing for us, but I'm trying to think of the api consumer instead of us.